src/hooks/useKeyboardNavigation.js

```diff
@@ -0,0 +1,64 @@
+import PropTypes from 'prop-types';
+
+import { ArrowDown, ArrowLeft, ArrowRight, ArrowUp } from './constants';
+
+const useKeyboardNavigation = (
+  categoriesNavRef,
+  emojiSearchRef,
+  emojiListRef
+) => {
```

Comment:
Like me, you're probably seeing a lot of repetition in this file. It makes sense, we're doing many similar things here.

There are many repeated imperative code here and calls to the same APIs. This can get error prone, and when we want to make a change in the future, we might have to change it in multiple places.

Sounds like using shared constants and functions might save us some hassle here.

For example:
1. Getting the activeElement at the entrance to each function.
2. writing a function `focusElement(element)` - less writing, and you can perform some safety checks within that function (does the element even exist?)
3. There are a few more, just try to think - what am I doing a couple of times here and can be "automated"?

Furthermore, even for stuff we're not doing more than once - we can reduce a lot of the complexity of each function, and improve its readability by replacing these imperative actions with declarative operations.

- `focusNextCategory()`
- `focusPrevCategory()`
- `focusNextSection()`
- `focusPrevSection()`
- `focusNextEmoji()`
- `focusPrevEmoji()`
- `focusSkinToneSelector()`
- `isInSearchBar()`
- `isInCategoriesNavigation()`
- `isInEmojisList()`
- `isInSkinToneSelector()`
- `isUp()`
- `isDown()`
- `isLeft()`
- `isRight()`

Each of these will probably use the `focusElement` we defined before, and will hide the "ugly" parts of your functions in easier to read and reason about calls.

Reply:
Hi.
Thinking out loud..
In order to implement the focusNextSection\focusPrevSection in a generic way, I should have an array with all the sections + a member that indicates the current one.
Every section will have its name just for the readability and a focus method.

Reply:
If you look closely at the list, I wrote quite a few functions that are still irrelevant at the moment, though I _do_ think it is still applicable.

The idea behind these functions is that we can hide the ugliness within them, so we have declarative functions that contain the imperative operations.

Inside  `focusNextSection` we can do all the gymnastics that we want (while still using the other functions that we defined) and "magically" find the correct one. Right now it can be done by - oh, we're in the search, let's go to the emojis, but in the future, it could use the more generic approach

Reply:
I've already implemented the generic approach,
let me know if it makes us problems, and then I will replace it with the KISS approach.